File: adaptor_test/node.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>

void callback(const std_msgs::String::ConstPtr& msg, const std::string& topic_name) {
    ROS_INFO("Received from %s: %s", topic_name.c_str(), msg->data.c_str());
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "multi_topic_subscriber");
    ros::NodeHandle nh;

    if (argc < 2) {
        ROS_ERROR("Usage: %s <n_topics>", argv[0]);
        return 1;
    }
    const int n_topics = std::atoi(argv[1]);

    std::vector<ros::Subscriber> subscribers;
    for (int i = 0; i < n_topics; ++i) {
        ROS_INFO("starting topic %d.", i);
        std::string topic = "/topic_" + std::to_string(i);
        subscribers.push_back(nh.subscribe<std_msgs::String>(topic, 10, boost::bind(callback, _1, topic)));
    }

    ros::spin();
    ROS_INFO("Spin complete.\n");
    return 0;
}

File: adaptor_test/node.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import String
import sys

def make_callback(topic_name):
    def callback(msg):
        rospy.loginfo(f"Received from {topic_name}: {msg.data}")
    return callback

if __name__ == "__main__":
    rospy.init_node("multi_topic_subscriber", anonymous=True)

    if len(sys.argv) < 2:
        rospy.logerr(f"Usage: {sys.argv[0]} <n_topics>")
        sys.exit(1)

    n_topics = int(sys.argv[1])
    subscribers = []

    for i in range(n_topics):
        rospy.loginfo(f"Starting topic {i}.")
        topic = f"/topic_{i}"
        subscribers.append(rospy.Subscriber(topic, String, make_callback(topic)))

    rospy.spin()
    rospy.loginfo("Spin complete.")
